### ISSUE TITLE
Android

### ISSUE THREAD

**Issue by weis** [Softmotions/ejdb#98]
We've managed to get ejdb running on Android using the android ndk.
Only a few small corrections were needed:

1) Android has no glob.c/glob.h so we use the one found on 
https://github.com/danstowell/GlastoCollider-Android/blob/master/jni/Source/libc/glob.c
2) two small patches were needed for stuff that isn't available on Android:

```
diff --git a/tcejdb/myconf.h b/tcejdb/myconf.h
index 11c7439..5b3b85a 100644
--- a/tcejdb/myconf.h
+++ b/tcejdb/myconf.h
@@ -286,7 +286,7 @@
    sizeof(TC_a) == sizeof(int16_t) ? _issigned(TC_a) ? INT16_MAX : UINT16_MAX : \
    _issigned(TC_a) ? INT8_MAX : UINT8_MAX))

-#if defined(_SYS_FREEBSD_) || defined(_SYS_NETBSD_) || defined(_SYS_OPENBSD_)
+#if defined(_SYS_FREEBSD_) || defined(_SYS_NETBSD_) || defined(_SYS_OPENBSD_) || defined (__ANDROID__)
 #define nan(TC_a)      strtod("nan", NULL)
 #define nanl(TC_a)     ((long double)strtod("nan", NULL))
 #endif
diff --git a/tcejdb/tcutil.c b/tcejdb/tcutil.c
index 9fe9cbf..98add30 100644
--- a/tcejdb/tcutil.c
+++ b/tcejdb/tcutil.c
@@ -5691,7 +5691,7 @@ int tcdayofweek(int year, int mon, int day) {
    `tm' specifies the pointer to the time structure.
    The return value is the GMT. */
 static time_t tcmkgmtime(struct tm *tm) {
-#if defined(_SYS_LINUX_)
+#if defined(_SYS_LINUX_) && !defined(__ANDROID__)
     assert(tm);
     return timegm(tm);
 #else
```

3) We decided to decrease the default size of the memory mapping, but I don't know if that is really neccessary:

```
diff --git a/tcejdb/tchdb.c b/tcejdb/tchdb.c
index b4e2632..5dd8c6b 100644
--- a/tcejdb/tchdb.c
+++ b/tcejdb/tchdb.c
@@ -47,6 +47,8 @@

 #ifdef _WIN32
 #define HDBDEFXMSIZ    _maxof(off_t)     // default size of the extra mapped memory
+#elif defined __ANDROID__
+#define HDBDEFXMSIZ    (8LL<<20)        // default size of the extra mapped memory
 #else
 #define HDBDEFXMSIZ    (64LL<<20)        // default size of the extra mapped memory
 #endif
diff --git a/tcejdb/tctdb.c b/tcejdb/tctdb.c
index a623d0a..0e480e3 100644
--- a/tcejdb/tctdb.c
+++ b/tcejdb/tctdb.c
@@ -29,7 +29,11 @@
 #define TDBDEFFPOW     10                // default free block pool power
 #define TDBDEFLCNUM    4096              // default number of leaf cache
 #define TDBDEFNCNUM    512               // default number of node cache
+#if defined __ANDROID__
+#define TDBDEFXMSIZ    (8LL<<20)        // default size of the extra mapped memory
+#else
 #define TDBDEFXMSIZ    (64LL<<20)        // default size of the extra mapped memory
+#endif
 #define TDBIDXSUFFIX   "idx"             // suffix of column index file
 #define TDBIDXLMEMB    64                // number of members in each leaf of the index
 #define TDBIDXNMEMB    256               // number of members in each node of the index
```

One thing that doesn't seem to work is using JBIDXNUM indexes. 
They work for $orderby clauses but if we use them in a normal search, 
we always get result count zero. We have not further investigated this issue.

**Comment by vladp** [comment]
+1. Also would like very much to see a working android port with easy to gradle-compatible binary distribution.

**Comment by adamansky** [comment]
Closed in favour to ejdb 2.0 https://github.com/Softmotions/ejdb/blob/master/README.md
